examples/vision/image_captioning.py

```diff
@@ -0,0 +1,509 @@
+    return cnn_model
+
+
+class TransformerEncoderBlock(layers.Layer):
+    def __init__(self, embed_dim, dense_dim, num_heads, **kwargs):
+        super().__init__(**kwargs)
+        self.embed_dim = embed_dim
+        self.dense_dim = dense_dim
+        self.num_heads = num_heads
+        self.attention = layers.MultiHeadAttention(
+            num_heads=num_heads, key_dim=embed_dim
+        )
+        self.dense_proj = layers.Dense(embed_dim, activation="relu")
+        self.layernorm_1 = layers.LayerNormalization()
+
+    def call(self, inputs, training, mask=None):
+        inputs = self.dense_proj(inputs)
+        attention_output = self.attention(
+            query=inputs, value=inputs, key=inputs, attention_mask=None
+        )
+        proj_input = self.layernorm_1(inputs + attention_output)
+        return proj_input
+
+
+class PositionalEmbedding(layers.Layer):
+    def __init__(self, sequence_length, vocab_size, embed_dim, **kwargs):
+        super().__init__(**kwargs)
+        self.token_embeddings = layers.Embedding(
+            input_dim=vocab_size, output_dim=embed_dim
+        )
+        self.position_embeddings = layers.Embedding(
+            input_dim=sequence_length, output_dim=embed_dim
+        )
+        self.sequence_length = sequence_length
+        self.vocab_size = vocab_size
+        self.embed_dim = embed_dim
+
+    def call(self, inputs):
+        length = tf.shape(inputs)[-1]
+        positions = tf.range(start=0, limit=length, delta=1)
+        embedded_tokens = self.token_embeddings(inputs)
+        embedded_positions = self.position_embeddings(positions)
+        return embedded_tokens + embedded_positions
+
+    def compute_mask(self, inputs, mask=None):
+        return tf.math.not_equal(inputs, 0)
+
+
+class TransformerDecoderBlock(layers.Layer):
+    def __init__(self, embed_dim, ff_dim, num_heads, **kwargs):
+        super().__init__(**kwargs)
+        self.embed_dim = embed_dim
+        self.ff_dim = ff_dim
+        self.num_heads = num_heads
+        self.attention_1 = layers.MultiHeadAttention(
+            num_heads=num_heads, key_dim=embed_dim
+        )
+        self.attention_2 = layers.MultiHeadAttention(
+            num_heads=num_heads, key_dim=embed_dim
+        )
+        self.dense_proj = keras.Sequential(
+            [layers.Dense(ff_dim, activation="relu"), layers.Dense(embed_dim)]
+        )
+        self.layernorm_1 = layers.LayerNormalization()
+        self.layernorm_2 = layers.LayerNormalization()
+        self.layernorm_3 = layers.LayerNormalization()
+
+        self.pe = PositionalEmbedding(
+            embed_dim=EMBED_DIM, sequence_length=SEQ_LENGTH, vocab_size=VOCAB_SIZE
+        )
+        self.out = layers.Dense(VOCAB_SIZE, activation="softmax")
+        self.dropout = layers.Dropout(0.5)
+
+    def call(self, inputs, encoder_outputs, training, mask=None):
+        inputs = self.pe(inputs)
+        causal_mask = self.get_causal_attention_mask(inputs)
```

Comment:
I think you should take the OR of the padding mask (`mask` argument) and the causal mask, no? Otherwise you will pay attention to padded locations.

Reply:
Ok. Let me update that

Reply:
What about this one? Because the encoder output is not padded at all, do we still need to use the `padding_mask`?

Reply:
The padding mask (`mask` argument) here refers to the `inputs` sequence which will need padding.

You'd also need to pass the `mask` argument here:

```
batch_seq_pred = self.decoder(batch_seq_inp, encoder_out, training=training)
```

And take the mask into account when computing the loss and accuracy as well.

Reply:
Both the causal mask and the padding mask refer to the same sequence (the caption), that's why they need to be combined

Reply:
Thanks @fchollet for the detailed feedback. Yeah, I see now that it is a mistake in the code. Apologies for that. Here is what I have figured out:

```python

# Pass this mask to the decoder block
mask = tf.cast(tf.math.not_equal(batch_seq_inp,  0), tf.int32)  # (batch_size, seq_len)

# Compute the causal mask in the attention block on PE output
causal_mask = get_causal_attention_mask(positional_embed_out) # (batch_size, seq_len, seq_len)

# Compute the padding mask
padding_mask = tf.cast(mask[:, tf.newaxis, :], tf.int32) # (batch_size, 1, seq_len)

# Take the minimum of two
padding_mask = tf.minimum(causal_mask, padding_mak) # (batch_size, seq_len, seq_len)

# This mask has to be passed to the attention layer where encoder output and the first attention layer
# output is combined
attention_output_2 = self.attention_2(
            query=out_1,
            value=encoder_outputs,
            key=encoder_outputs,
            attention_mask=padding_mask,
        )
```

But there is a problem
```
Shape of query (out_1) -> (B, 39, 512)
Shape of value (encoder_out) -> (B, 100, 512)
Shape of padding mask - > (B, 39, 39)
```

For the attention layer to work as expected, the padding mask should have a shape of `(B, 39, 100)`. Should we pad the padding_mask to make sure it has the correct shape or do you have another solution to propose?

Reply:
The minimum mask (combined) has to be passed to `attention_1`.

As for `attention_2`... for a given sample, `mask[i, j]` indicates the interaction between `caption[i]` and `image[j]`. We want that to be 0 for `i > caption_length` so you need to pass the padding mask. I think shape `(B, 39, 1)` should work. No need to pass the combined mask at that step, I think, because that's only relevant when the caption interacts with itself, no when the caption interacts with the image.